File: 10000/15486.cpp
#include<iostream>
using namespace std;

int n;
int a[1500001][2];
int dp[1500001];

void input() {
	cin >> n;
	for (int i = 1; i <= n; ++i) {
		cin >> a[i][0] >> a[i][1];
	}
}

void solve() {
	input();
	int answer = 0;
	for (int i = 1; i <= n; ++i) {
		if (answer < dp[i]) answer = dp[i];
		int day = i + a[i][0];
		dp[day] = max(dp[day], answer + a[i][1]);
		if (dp[i] > answer) answer = dp[i];
	}
	if (dp[n + 1] > answer) answer = dp[n + 1];
	cout << answer;
}

int main() {
	cin.tie(0), cout.tie(0);
	ios_base::sync_with_stdio(0);
	solve();
	return 0;
}

File: 10000/15486.java
import java.util.*;
import java.io.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int n;
    static int[] dp;
    static int[][] councils;

    static void input() throws IOException {
        n = Integer.parseInt(br.readLine());
        councils = new int[n + 2][2];
        dp = new int[n + 2];
        for (int i = 1; i <= n; ++i) {
            st = new StringTokenizer(br.readLine());
            councils[i][0] = Integer.parseInt(st.nextToken());
            councils[i][1] = Integer.parseInt(st.nextToken());
        }
    }

    static void solution() {
        int answer = 0;
        for (int i = 1; i <= n + 1; ++i) {
            answer = Math.max(answer, dp[i]);
            if(councils[i][0] + i > n + 1) continue;
            dp[i + councils[i][0]] = Math.max(dp[i + councils[i][0]], answer + councils[i][1]);
        }
        System.out.println(answer);
    }

    public static void main(String[] args) throws IOException {
        input();
        solution();
    }
}
